### PR TITLE
Feature/268 add enums for network config settings

### PR DESCRIPTION
**Pull request by ArdtK**

### CODE REVIEW

**Review by Carsopre — CHANGES_REQUESTED** [review]

File: ra2ce/graph/network_config_data/network_config_data_validator.py [hunk 1]
```diff
@@ -89,11 +90,26 @@ def validate(self) -> ValidationReport:
...
+        _accepted_values = ", ".join(
+            [member.config_value for member in type(enum)][:-1]
+        )
+        return (
+            f"Wrong input to property [ {key} ], has to be one of: {_accepted_values}."
+        )
+
+    def _validate_enum(self, enum: Ra2ceEnumBase) -> ValidationReport:
+        _report = ValidationReport()
+        if enum.name == "INVALID":
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/graph/network_config_data/network_config_data_validator.py:109]: The check needs to be done on the base to make it applicable for other enums. The base can't contain enums, which gives an linting error.
Adjusted though.

File: ra2ce/graph/network_config_data/network_config_data_reader.py [hunk 2]
```diff
@@ -138,9 +140,15 @@ def _get_path_list(
...
 
+    def _get_enum(
+        self, section_name: str, property: str, enum_type: type(Enum)
+    ) -> Enum:
+        return enum_type.get_enum(self._parser.get(section_name, property))
+
     def get_network_section(self) -> NetworkSection:
         _section = "network"
         _network_section = NetworkSection(**self._parser[_section])
+        _network_section.source = self._get_enum(_section, "source", SourceEnum)
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/graph/network_config_data/network_config_data_reader.py:151]: Fair point.

File: tests/test_data/acceptance_test_data/output/network.ini
[the remark this answers is not in the document]
Comment (ArdtK) [reply tests/test_data/acceptance_test_data/output/network.ini:1]: Correct. Will revert. Done unintentionally.

File: ra2ce/graph/network_config_data/network_config_data_reader.py [hunk 3]
```diff
@@ -138,9 +140,15 @@ def _get_path_list(
         )
         return list(map(self._get_str_as_path, _value_list))
 
+    def _get_enum(
+        self, section_name: str, property: str, enum_type: type(Enum)
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/graph/network_config_data/network_config_data_reader.py:144]: This works.

[the remark this answers is not in the document]
Comment (Carsopre) [reply ra2ce/graph/network_config_data/network_config_data_reader.py]: Right, also something that got updated, however, you need to use the brackets ( `type[Enum]`), not the parenthesis, if that's not allowed, please revert back to `Type`. More info: https://docs.python.org/3/library/typing.html#the-type-of-class-objects

File: ra2ce/graph/network_config_data/network_config_data_reader.py [hunk 4]
```diff
@@ -153,8 +160,14 @@ def get_network_section(self) -> NetworkSection:
...
         )
-        _network_section.road_types = self._parser.getlist(
-            _section, "road_types", fallback=_network_section.road_types
+        _network_section.network_type = NetworkTypeEnum.get_enum(
+            self._parser.get(_section, "network_type", fallback=None)
+        )
+        _network_section.road_types = list(
+            map(
+                RoadTypeEnum.get_enum,
+                self._parser.getlist(_section, "road_types", fallback=[None]),
```
Comment (Carsopre) [inline ra2ce/graph/network_config_data/network_config_data_reader.py:169]: The default value should be `[]`?
```suggestion
                self._parser.getlist(_section, "road_types", fallback=[]),
```

**Review by Carsopre — CHANGES_REQUESTED** [review]
I'm still unsure about the `NONE` enum, if a list does not contain elements it should just simply be empty.
Aside from this remark, there are other minor ones. Of course TC needs to be green.

File: ra2ce/analyses/analysis_config_data/enums/weighing_enum.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+from ra2ce.configuration.ra2ce_enum_base import Ra2ceEnumBase
+
+
+class WeighingEnum(Ra2ceEnumBase):
+    NONE = 0
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/analyses/analysis_config_data/enums/weighing_enum.py:5]: Added in Ra2ceEnumBase.

File: ra2ce/analyses/indirect/analyses_indirect.py [hunk 6]
```diff
@@ -1071,9 +1076,9 @@ def _save_gpkg_analysis(
                 )
                 graph_to_gpkg(base_graph, gpkg_path_edges, gpkg_path_nodes)
 
-            if analysis.weighing == "distance":
+            if analysis.weighing == WeighingEnum.DISTANCE:
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/analyses/indirect/analyses_indirect.py:1079]: Check

File: ra2ce/configuration/ra2ce_enum_base.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        Args:
+            key (str): Enum key (name)
+
+        Returns:
+            bool: If the given key is not a valid key
+        """
+        if self.name == "INVALID":
+            return False
+        else:
+            return True
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/configuration/ra2ce_enum_base.py:42]: Agree

File: ra2ce/graph/network_config_data/enums/aggregate_wl_enum.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+from ra2ce.configuration.ra2ce_enum_base import Ra2ceEnumBase
+
+
+class AggregateWlEnum(Ra2ceEnumBase):
+    NONE = 0
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/graph/network_config_data/enums/aggregate_wl_enum.py:5]: Explained

File: ra2ce/graph/network_config_data/enums/road_type_enum.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    TRUNK = 3
+    TRUNK_LINK = 4
+    PRIMARY = 5
+    PRIMARY_LINK = 6
+    SECONDARY = 7
+    SECONDARY_LINK = 8
+    TERTIARY = 9
+    TERTIARY_LINK = 10
+    UNCLASSIFIED = 11
+    RESIDENTIAL = 12
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/graph/network_config_data/enums/road_type_enum.py:17]: Chose 98

File: ra2ce/graph/network_config_data/network_config_data.py [hunk 10]
```diff
@@ -38,13 +42,13 @@ class ProjectSection:
...
-    source: str = ""  # should be enum
+    source: SourceEnum = field(default_factory=lambda: SourceEnum.INVALID)
     primary_file: list[Path] = field(default_factory=list)
     diversion_file: list[Path] = field(default_factory=list)
     file_id: str = ""
     polygon: Optional[Path] = None
-    network_type: str = ""  # Should be enum
-    road_types: list[str] = field(default_factory=list)
+    network_type: NetworkTypeEnum = field(default_factory=lambda: NetworkTypeEnum.NONE)
+    road_types: list[RoadTypeEnum] = field(default_factory=lambda: [RoadTypeEnum.NONE])
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply ra2ce/graph/network_config_data/network_config_data.py:51]: Adjusted

File: ra2ce/graph/network_config_data/network_config_data_reader.py (hunk 4 above)
Comment (ArdtK) [reply ra2ce/graph/network_config_data/network_config_data_reader.py:169]: Adjusted

**Review by Carsopre — APPROVED** [review]
Approved although the parsing of the `weighing_enum.py` can be improved so to reduce its defined enums.

**Pull request merged**